Hive metastore integration with installation files and a sample application

srikumar003 opened the pull request.

YiannisGkoufas commented on release-tools/Makefile:
```diff
@@ -102,7 +102,7 @@ ifeq ($(PULL_COMPONENTS), true)
 else
 	$(call install_local,csi-s3,$(CSI_S3_IMAGE))
-	$(call install_local,csi-driver-nfs,$(CSI_NFS_IMAGE))
+#	$(call install_local,csi-driver-nfs,$(CSI_NFS_IMAGE))
```
Why is this commented out? it doesn't build the nfs driver. if I comment it out it fails to build.
```
go: cannot determine module path for source directory /csi-driver-nfs (outside GOPATH, no import comments)
release-tools/build.make:64: recipe for target 'build-nfsplugin' failed
make: *** [build-nfsplugin] Error 1
make: *** [build-containers] Error 2
```

srikumar003 commented on src/dataset-operator/pkg/controller/dataset/dataset_controller.go:
```diff
@@ -428,59 +428,95 @@ func processRemoteDataset(cr *comv1alpha1.Dataset, rc *ReconcileDataset) (reconc
 
-			if err = controllerutil.SetControllerReference(cr, configMapObject, rc.scheme); err != nil {
+		if err != nil && errors.IsNotFound(err) {
+			processRemoteDatasetLogger.Info("Creating new configMap", "configMap.namespace",
```
Fixed it in https://github.com/srikumar003/dataset-lifecycle-framework/commit/6a9b2abb10263fd88841c452c938fdaa5086d79b

christian-pinto replied:
yep, fix works. Just a minor stylish thing to be added, added a comment for this.

christian-pinto commented on Makefile:
```diff
@@ -1,5 +1,5 @@
-DOCKER_REGISTRY_COMPONENTS := the_registry_to_use_for_components
-DOCKER_REGISTRY_SECRET := your_already_installed_secrets
+DOCKER_REGISTRY_COMPONENTS := res-drl-hpc-docker-local.artifactory.swg-devops.com
+DOCKER_REGISTRY_SECRET := art-drl-hpc
```
please revert these changes

srikumar003 replied:
Fixed this in 9c66ac156452e7d4c3fd366a85116342b694b03f

christian-pinto commented on Makefile:
```diff
@@ -8,8 +8,8 @@ DOCKER_REGISTRY_SIDECARS := quay.io/k8scsi
-DATASET_OPERATOR_NAMESPACE := default
-NAMESPACES_TO_MONITOR := default
+DATASET_OPERATOR_NAMESPACE := drlmcdlf
+NAMESPACES_TO_MONITOR := drlmcdlf
```
And these too :smiley:

srikumar003 replied:
fixed 9c66ac156452e7d4c3fd366a85116342b694b03f

christian-pinto commented on src/dataset-operator/pkg/admissioncontroller/mutate.go:
```diff
@@ -127,53 +132,49 @@ func Mutate(body []byte) ([]byte, error) {
+				//TODO: Check if the configmap reference exists in the API server
+
+				configmap_ref := map[string]interface{}{
+					"prefix": config_toinject,
```
maybe add a "-" after the prefix otherwise env variables names look weird such as **example-datasetbucket**. Please do the same with the secret

srikumar003 commented on examples/hive/k8s/README.md:
````diff
@@ -0,0 +1,37 @@
+if you are using the Minio install as described in the main installation guide, the following applies.
+
+```
+$ export S3_ENDPOINT=$(minikube service minio-service --url)
````
Changed Hive Readme to use Nooba b1ceb62099b1b67a088c682d361c3bc3a16c3896

srikumar003 commented on examples/hive/k8s/README.md:
````diff
@@ -0,0 +1,37 @@
+$ envsubst < conf/metastore-site.tmpl | tee conf/metastore-site.xml
+```
+```
+$ mv deploy/s3-secret.tmpl deploy/s3-secret.yaml
````
No need for this file 2fb204cc047990efb1406b068409e8952bca34c9

srikumar003 commented on examples/hive/k8s/README.md:
````diff
@@ -0,0 +1,37 @@
+```
+```
+$ mv deploy/s3-secret.tmpl deploy/s3-secret.yaml
+$ kubectl apply -f deploy/s3-secret.yaml
````
Updated Readme to use these variables b1ceb62099b1b67a088c682d361c3bc3a16c3896

srikumar003 commented on examples/hive/k8s/Makefile:
```diff
@@ -57,7 +65,11 @@ push-images: build-images
+	$(call load_containers_minikube, $(HIVEMETASTORE_IMAGE), hivemetastore)
+	$(call load_containers_minikube, $(HIVESERVER_IMAGE), hiveserver)
+
+deploy-hive: push-images
```
Fixed this in 9c66ac156452e7d4c3fd366a85116342b694b03f

srikumar003 commented on examples/hive/k8s/Makefile:
```diff
@@ -32,6 +34,12 @@ MAKE_ENV += HIVEMETASTORE_IMAGE
 
 #K8S_FILES += $(shell find $(HIVE_FILE_PATH)/deploy -maxdepth 1 -name '*.yaml')
+define load_containers_minikube
+    @docker save $(2) | gzip > _tmp/$(1).tar.gz;\
```
Fixed this in 9c66ac156452e7d4c3fd366a85116342b694b03f

srikumar003 commented on examples/hive/k8s/Makefile:
```diff
@@ -57,7 +65,11 @@ push-images: build-images
-deploy-hive: push-images 
+minikube-load-containers: build-images
+	$(call load_containers_minikube, $(HIVEMETASTORE_IMAGE), hivemetastore)
+	$(call load_containers_minikube, $(HIVESERVER_IMAGE), hiveserver)
```
Swapped the order of parameters 1 and 2 in the function 9c66ac156452e7d4c3fd366a85116342b694b03f

christian-pinto commented on examples/hive/k8s/deploy/hiveserver.yaml:
```diff
@@ -0,0 +1,46 @@
+      containers:
+        - name: hiveserver-container
+          image: ${HIVESERVER_IMAGE}
+          imagePullPolicy: "Always"
```
Change pull policy, it fails in minikube because it tries pulling the image. **IfNotPresent** should make the trick

christian-pinto commented on examples/hive/k8s/deploy/metastore.yaml:
```diff
@@ -0,0 +1,40 @@
+      containers:
+        - name: hivemetastore-container
+          image: ${HIVEMETASTORE_IMAGE}
+          imagePullPolicy: "Always"
```
Change pull policy, it fails in minikube because it tries pulling the image. **IfNotPresent** should make the trick

christian-pinto commented on examples/hive/sampleapp/populate.sh:
```diff
@@ -0,0 +1,13 @@
+DOCKER_REGISTRY_COMPONENTS="registry_used_in_installation"
+HIVESERVER_IMAGE="hive-server:latest"
+
+docker run -v $(PWD):/sampleapp -it ${DOCKER_REGISTRY_COMPONENTS}/${HIVESERVER_IMAGE} bin/beeline -u "jdbc:hive2://$HIVE_CLI_IP:$HIVE_CLI_PORT/;transportMode=http;httpPath=/cliservice" -f /sampleapp/sample.hql
```
$(PWD) -> ${PWD}

christian-pinto replied:
Also, this last command returns:
```
20/03/13 09:26:22 [main]: WARN jdbc.HiveConnection: Failed to connect to 192.168.39.51:31321
Error: Could not open client transport with JDBC Uri: jdbc:hive2://192.168.39.51:31321/;transportMode=http;httpPath=/cliservice: Could not establish connection to jdbc:hive2://192.168.39.51:31321/;transportMode=http;httpPath=/cliservice: org.apache.http.conn.HttpHostConnectException: Connect to 192.168.39.51:31321 [/192.168.39.51] failed: Connection refused (Connection refused) (state=08S01,code=0)

```

srikumar003 replied:
Variable substitution fixed in https://github.com/srikumar003/dataset-lifecycle-framework/commit/74d2c763bc1bdb3190430c2ea89f60c7c6cfc999

Provided a convenience script for setting up sample app

christian-pinto commented on examples/hive/sampleapp/README.md:
````diff
@@ -1,23 +1,45 @@
 
 ```
-$ mc cp books.csv local/book-test
+$ s3cmd put books.csv s3://book-test/
````
All this could be automated the same way we load example data in NooBaa for testing the regular datasets. There is a docker image wtih _awscli_ that is built during the installation process and you could use that one to load the csv inside the new bucket.

See [data-loader-noobaa.yaml](https://github.com/IBM/dataset-lifecycle-framework/blob/master/examples/noobaa/data-loader-noobaa.yaml) and [noobaa_install.sh](https://github.com/IBM/dataset-lifecycle-framework/blob/master/examples/noobaa/noobaa_install.sh) functions `build_data_loader` and `run_data_loader`.

srikumar003 replied:
Provided a convenience script for creating the sample application: https://github.com/srikumar003/dataset-lifecycle-framework/commit/ae9a885eed1c985fabbad5155531a2ea7734cb03

YiannisGkoufas commented:
I am probably configuring something wrong in the hive example on the `s3cmd --configure` step.
While this `s3cmd ls` works and shows the buckets:
```
2020-03-23 10:48  s3://first.bucket
2020-03-23 10:48  s3://my-bucket-5fc7f163-aeba-459b-8b66-69caeeec3887
```
However, when I try to do this `s3cmd mb s3://book-test` I get the error:
`ERROR: S3 error: 403 (InvalidAccessKeyId): The AWS Access Key Id you provided does not exist in our records.`

As I understand the `s3cmd --configure` command creates a `~/.s3cfg` file which stores the configuration parameters. Maybe showing an example of `~/.s3cfg` might be helpful

srikumar003 commented on examples/hive/k8s/README.md:
````diff
@@ -0,0 +1,37 @@
+$ kubectl apply -f deploy/s3-secret.yaml
+```
+```
+$ kubectl apply -f deploy/postgres-secret.yaml
````
The Makefile for hive installation now takes care of all the configuration file modification. Fixed in https://github.com/srikumar003/dataset-lifecycle-framework/commit/2f40baf08e9f33f466e8c10283cf3cf4ce885ac4

srikumar003 commented:
> I am probably configuring something wrong in the hive example on the `s3cmd --configure` step.
> While this `s3cmd ls` works and shows the buckets:
> 
> ```
> 2020-03-23 10:48  s3://first.bucket
> 2020-03-23 10:48  s3://my-bucket-5fc7f163-aeba-459b-8b66-69caeeec3887
> ```
> 
> However, when I try to do this `s3cmd mb s3://book-test` I get the error:
> `ERROR: S3 error: 403 (InvalidAccessKeyId): The AWS Access Key Id you provided does not exist in our records.`
> 
> As I understand the `s3cmd --configure` command creates a `~/.s3cfg` file which stores the configuration parameters. Maybe showing an example of `~/.s3cfg` might be helpful

Removed the need for configuring `s3cmd`. We are using the Docker image for awscli that is wrapped in a convenience script to set up the sample application. Fixed in https://github.com/srikumar003/dataset-lifecycle-framework/commit/ae9a885eed1c985fabbad5155531a2ea7734cb03

christian-pinto commented:
There is still an issue with the main makefile. Your branch contains the wrong values for registry, secret  and namespace that make the process fail while installing on minikube.

christian-pinto commented:
Looks good now!
